### PR TITLE
RATEST-299: E2E Test: Person feature workflow

### PR DESCRIPTION
**Pull request by mherman22**
**Ticket ID**: https://issues.openmrs.org/browse/RATEST-299

**Goal:**
To Combine selenium person tests linked to this ticket and rewrite in cucumber framework to form an end to end workflow.
The selenium tests in context include;

`CreatePersonTest`
`EditPersonTest`
`RetirePersonTest`
`DeletePersonTest.`

**what i did to achieve the goal**

Created a feature file - `person.feature`
wrote step definitions - `PersonSteps.java`
Completed the test in Selenium using the OMRS QA Framework
Added a "Person" test status pill to the QA Framework Repo README under the 2.x RefApp section

### CODE REVIEW

**Comment by mherman22** [comment]
cc: @kdaud @sherrif10

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AddPersonPage.java [hunk 1]
```diff
@@ -14,23 +14,23 @@
 
 public class AddPersonPage extends Page {
 
-	private static final By PERSON_NAME = By.id("personName");
-	private static final By BIRTHDATE = By.id("birthdate");
+	private static final By PERSON_NAME_FIELD = By.id("personName");
+	private static final By BIRTHDATE = By.className("hasDatepicker");
```
Comment (kdaud) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AddPersonPage.java:18]: is `BIRTHDAY` a button, filed, link or what?

File: qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AddPersonPage.java [hunk 2]
```diff
@@ -14,31 +14,31 @@
 
 public class AddPersonPage extends Page {
 
-	private static final By PERSON_NAME = By.id("personName");
-	private static final By BIRTHDATE = By.id("birthdate");
-	private static final By AGE = By.id("age");
+	private static final By PERSON_NAME_FIELD = By.id("personName");
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AddPersonPage.java:17]: Please use the default element attribute description
```suggestion
	private static final By PERSON_NAME= By.id("personName");
	private static final By BIRTHDATE = By.id("birthdate");
	private static final By AGE = By.id("age");
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/main/java/org/openmrs/contrib/qaframework/page/AddPersonPage.java:17]: Those are more readable than the current changes.

**Comment by mherman22** [comment]
cc: @kdaud @sherrif10

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    public void launchSystemAdministrationPage() {
+    	administrationPage = homePage.goToAdministration();
+    }
+    
+    @When ("User clicks the manage Persons link on the administration page")
+    public void launchAdvancedAdministrationPage() {
+    	managePersonPage = administrationPage.clickOnManagePersons();
+    }
+    
+    @And ("User clicks the create person link on the manage persons page")
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:66]:
```suggestion
    @And ("User clicks on  the create person link on the manage persons page") Looks more readable. Update all the feature files with user **clicks on the** . where neccessary.
```

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:66]: @mherman22  What we want to ephasize is user readability. Take in mind that, if these feature files get rendered in cucumber, it helps a new user to clearly tell what the test is functioning, so we need to clealy use a most readable scenarioa. This is yet to be merged, some slight changes needed.

Comment (mherman22) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:66]: Alright , I get the message. Look out for my next commit!!!!

**Review by sherrif10 — APPROVED** [review]
Thanks, @mherman22 for the great work. @kdaud Looking forward for final reviews.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    public void setRetireReason() {
+    	personFormPage.setRetireReason("retire reason");
+    }
+    
+    @Then ("User clicks on the retire person button")
+    public void retirePerson() {
+    	personFormPage.retirePerson();
+    }
+}
+
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:115]: isn't one empty line at the EOF sufficient enough to cater for git collaborators?

**Comment by mherman22** [comment]
cc: @kdaud

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    private PersonFormPage personFormPage;
+    private AddPersonPage addPersonPage;
+    private String personUuid;
+    private TestData.PersonInfo personInfo;
+    private final String PERSON_NAME = RandomStringUtils.randomAlphabetic(8);
+    private final String PERSON_AGE = RandomStringUtils.randomNumeric(2);
+    private final String PERSON_FAMILY_NAME = RandomStringUtils.randomAlphanumeric(8);
+	
+    @Before(RunTest.HOOK.SELENIUM_PERSON)
+    public void setup() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:44]: align with the method naming convention.

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/Person/person.feature [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    And a user clicks on the first person found from the search results
+    Then a user clicks on the delete person forever button
+
+  @selenium
+  @person
+  Scenario: Retiring a Person
+    When a user searches for the person
+    And a user clicks on the first person found from the search results
+    And a user fills in the reason for retiring person
+    Then a user clicks on the retire person button
```
Comment (sherrif10) [inline qaframework-bdd-tests/src/test/resources/features/refapp-2.x/Person/person.feature:38]: This looks good to be merged.One slight change. Add one line at the end of the file to cater for future git use cases.

**Review by sherrif10 — APPROVED** [review]
LGTM  cc @kdaud  More reviews and we get this guy merged

File: qaframework-bdd-tests/src/test/resources/features/refapp-2.x/Person/person.feature (hunk 6 above)
Comment (mherman22) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/Person/person.feature:38]: The line is there. Its just invisible

Comment (sherrif10) [reply qaframework-bdd-tests/src/test/resources/features/refapp-2.x/Person/person.feature:38]: Not quite certain why when editing is when the line gets visible. Thank you.Meanwhile let's wait for @kdaud final reviews

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	
+    @Before(RunTest.HOOK.SELENIUM_PERSON)
+    public void setUp() {
+	initiateWithLogin();
+	personInfo = TestData.generateRandomPerson();
+	personUuid = TestData.createPerson(personInfo);
+    }
+	
+    @After(RunTest.HOOK.SELENIUM_PERSON)
+    public void teardown() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:52]: align with the method naming convention.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    }
+	
+    @Given ("a user clicks on the system administration link on the home page")
+    public void launchSystemAdministrationPage() {
+    	administrationPage = homePage.goToAdministration();
+    }
+    
+    @Then ("the system loads the administration page")
+    public void loadAdministrationPage() {
+    	assertTrue(textExists("Administration"));
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:64]: which text appears when the system loads the System Administration page?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    }
+    
+    @And ("a user clicks on the manage Persons link on the administration page")
+    public void launchAdvancedAdministrationPage() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:68]: take a close look at the method name and find out whether it reflects what the method is doing.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java (hunk 8 above)
Comment (mherman22) [reply qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:64]: when `administrationPage = homePage.goToAdministration();` is triggered,  the test expects to land on the `Administration` page even when it goies through the `System Administration` page. 

have a look-see at this function 
```
	public AdministrationPage goToAdministration() {
		clickOn(SYSTEM_ADMINISTRATION);
		return new SystemAdministrationPage(this).goToAdvancedAdministration();
	}
```

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java [hunk 10]
```diff
@@ -28,6 +28,8 @@
 		public static final String CONDITION = "@condition";
 		public static final String SELENIUM = "@selenium";
 		public static final String FORM = "@form";
+		public static final String PERSON = "@person";
+		public static final String DASHBOARD = "@dashboard";
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/RunTest.java:32]: any reason for adding `DASHBOARD` hook in RunTest,java file?

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    	addPersonPage = managePersonPage.createPerson();
+    }
+    
+    @Then ("the system loads the create person page")
+    public void loadPersonFormPage() {
+    	assertTrue(textExists("Adding a Person"));
+    }
+
+    @And ("a user fills the create person form")
+    public void createPersonForm() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:88]: take a close look at the method name and find out whether it reflects what the method is doing.

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    	assertTrue(textExists("Person"));
+    }
+    
+    @When ("a user clicks on the create person link on the manage persons page")
+    public void clickOnCreatePersonLink() {
+    	addPersonPage = managePersonPage.createPerson();
+    }
+    
+    @Then ("the system loads the create person page")
+    public void loadPersonFormPage() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:83]: take a close look at the method and figure out whether its a form or a page!

File: qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    	assertThat(personFormPage.getFamilyName(), is("newFamilyName"));
+    }
+    
+    @When ("a user searches for the person")
+    public void searchForPersonToEdit() {
+    	managePersonPage.setPersonName(personInfo.givenName);
+    }
+    
+    @And ("a user clicks on the first person found from the search results")
+    public void clickFirstPersonFound() {
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/java/org/openmrs/contrib/qaframework/automation/PersonSteps.java:113]: what do you think of `clickFirstPersonFound()` vs `clickOnFirstPerson()` ?

**Review by kdaud — APPROVED** [review]

**Pull request merged**